dev/tools/pre-commit

```diff
@@ -73,3 +73,12 @@ then
     # clean up temporary files
     rm "$index" "$tree" "$fixed_index"
 fi
+
+# Check and fix formatting
+if command -v ocamlformat > /dev/null
+then
+    ocamlformat -i $(find plugins/micromega/*.ml{,i} ! -name micromega.ml)
+    git add -u
```

Comment:
Unrestricted `git add -u` is no good
Possibly instead we should call ocamlformat next to `check-eof-newline.sh --fix` (and so call it twice)
We can blindly call it on all modified .ml(i) files, the ocamlformat-ignore will restrict effects appropriately.

Reply:
I tried to apply your suggestion but did not manage to reach a working state. Have a look at my attempted patch:

```diff
diff --git a/dev/tools/pre-commit b/dev/tools/pre-commit
index 028269b19c..feb2e9dd00 100755
--- a/dev/tools/pre-commit
+++ b/dev/tools/pre-commit
@@ -7,7 +7,17 @@ set -e
 
 dev/tools/check-overlays.sh
 
+# Can we check and fix formatting?
+if ! command -v ocamlformat > /dev/null
+then
+    1>&2 echo "Warning: ocamlformat is not in path. Cannot check formatting."
+    function ocamlformat() {
+        :
+    }
+fi
+
 if ! git diff --cached --name-only -z | xargs -0 dev/tools/check-eof-newline.sh ||
+        ! git diff --cached --name-only -z | grep -E '.*\.mli?$' | xargs -0 ocamlformat --check ||
         ! git diff-index --check --cached HEAD >/dev/null 2>&1 ;
 then
     1>&2 echo "Auto fixing whitespace issues..."
@@ -36,6 +46,7 @@ then
     git apply --cached --whitespace=fix "$index"
     git apply --whitespace=fix "$index" 2>/dev/null # no need to repeat yourself
     git diff --cached --name-only -z | xargs -0 dev/tools/check-eof-newline.sh --fix
+    git diff --cached --name-only -z | grep -E '.*\.mli?$' | xargs -0 ocamlformat -i
     git add -u
     1>&2 echo #newline
 
@@ -73,12 +84,3 @@ then
     # clean up temporary files
     rm "$index" "$tree" "$fixed_index"
 fi
-
-# Check and fix formatting
-if command -v ocamlformat > /dev/null
-then
-    ocamlformat -i $(find plugins/micromega/*.ml{,i} ! -name micromega.ml)
-    git add -u
-else
-    1>&2 echo "Warning: ocamlformat is not in path. Could not check formatting."
-fi
```

Reply:
You probably need -z in your greps
Also you can't use bash functions with xargs, instead do
~~~bash
ocamlformat=$(command -v ocamlformat || echo true)
~~~
then call `xargs -0 ocamlformat ...`

Reply:
Thanks for your advice. This seems to have worked (see new commit). But there is something weird which is that formatting issues are restored after committing (even if they are fixed in the commit).